README.md

<!--- README.md is generated from README.Rmd. Please edit that file -->
prepairr: Interface to the automatic geometry repair tool 'prepair'
-------------------------------------------------------------------

[![lifecycle](https://img.shields.io/badge/Lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental) [![Travis Build Status](https://img.shields.io/travis/jeffreyhanson/prepairr/master.svg?label=Status)](https://travis-ci.org/jeffreyhanson/prepairr) [![Coverage Status](https://codecov.io/github/jeffreyhanson/prepairr/coverage.svg?branch=master)](https://codecov.io/github/jeffreyhanson/prepairr?branch=master)

### Overview

The *prepairr R* package provides an interface to the [*prepair* tool](https://github.com/tudelft3d/prepair) for repairing invalid polygon geometries (Ledoux, Arroyo Ohori & Meijers, 2014). This project was developed as an experiment to benchmark the performance of the *prepair* tool for repairing invalid geometries in large spatial datasets. Unfortunately, this package is actually **slower** than standard geometry cleaning tools (such as the [*lwgeom R* package](https://CRAN.R-project.org/package=lwgeom)). As a consequence, this package is unlikely to receive further development.

### Installation

The developmental version of the *prepairr R* package can be installed using the following R code. Please note that this package will only work on Unix systems and requires the [*sf R*](https://github.com/r-spatial/sf) package which requires additional software to be installed. If you encounter problems installing the *prepairr R* package, please consult the installation instructions for the [*sf R*](https://github.com/r-spatial/sf) package.

``` r
if (!require(remotes))
  remotes::install_github("jeffreyhanson/prepairr")
```

### Usage

Here we will explore the usage and (substandard) performance of this package. To achieve this, we will first load the *prepair* package and some other packages for working with spatial vector data.

``` r
# load packages
library(prepairr)
library(lwgeom)
library(rworldxtra)
```

Next, we will obtain a small spatial dataset containing invalid geometries. Specifically, we will import data delineating country boundaries (originally from [Natural Earth](https://www.naturalearthdata.com/)).

``` r
# import data, subset data to reduce total run time, and convert to sf format
data(countriesHigh, package = "rworldxtra")
x <- as(countriesHigh[c(12, 41, 132, 174, 190, 197, 211, 225), "ADMIN"], "sf")

# preview data
print(x)
```

    ## Simple feature collection with 8 features and 1 field
    ## geometry type:  MULTIPOLYGON
    ## dimension:      XY
    ## bbox:           xmin: -179.999 ymin: -89.99983 xmax: 180 ymax: 49.60178
    ## epsg (SRID):    4326
    ## proj4string:    +proj=longlat +datum=WGS84 +no_defs
    ##           ADMIN                       geometry
    ## 12   Antarctica MULTIPOLYGON (((-61.03006 -...
    ## 41        Chile MULTIPOLYGON (((-69.47578 -...
    ## 132     Liberia MULTIPOLYGON (((-9.433115 7...
    ## 174 New Zealand MULTIPOLYGON (((174.0249 -4...
    ## 190     Romania MULTIPOLYGON (((23.32512 43...
    ## 197     Senegal MULTIPOLYGON (((-13.08374 1...
    ## 211    Slovakia MULTIPOLYGON (((19.62695 49...
    ## 225       Tonga MULTIPOLYGON (((-175.3144 -...

``` r
# plot data
plot(st_geometry(x), main = "raw data")
```

<img src="man/figures/README-raw_data-1.png" style="display: block; margin: auto;" />

``` r
# verify that the dataset has invalid geometries
all(suppressWarnings(st_is_valid(x)))
```

    ## [1] FALSE

Now we will repair the data using the [*prepair* tool](https://github.com/tudelft3d/prepair). We will also record how long it takes to repair the data so we can compare it with standard data cleaning tools later. Please note that the timings reported below may be different on your system, since I have added `CXX11FLAGS += -O3 -march=native` to the Makevars file on my computer to increase performance of compiled code in R packages.

``` r
# repair geometries using st_prepair and record processing time
prepair_time <- system.time({
  y <- st_prepair(x)
})[[3]]

# preview cleaned data
print(y)
```

    ## Simple feature collection with 8 features and 1 field
    ## geometry type:  GEOMETRY
    ## dimension:      XY
    ## bbox:           xmin: -179.999 ymin: -89.99983 xmax: 180 ymax: 49.60178
    ## epsg (SRID):    4326
    ## proj4string:    +proj=longlat +datum=WGS84 +no_defs
    ##           ADMIN                       geometry
    ## 12   Antarctica MULTIPOLYGON (((-52.51818 -...
    ## 41        Chile MULTIPOLYGON (((-68.63828 -...
    ## 132     Liberia POLYGON ((-9.433114 7.99331...
    ## 174 New Zealand MULTIPOLYGON (((174.0249 -4...
    ## 190     Romania POLYGON ((23.19696 43.86275...
    ## 197     Senegal POLYGON ((-13.09161 15.5023...
    ## 211    Slovakia POLYGON ((19.62873 49.41132...
    ## 225       Tonga MULTIPOLYGON (((-175.3638 -...

``` r
# verify that all geometries are now valid,
# except they aren't so this tool is slow AND doesn't fix all geometry issues
all(st_is_valid(y))
```

    ## [1] FALSE

``` r
# print processing time with st_prepair (seconds)
print(prepair_time)
```

    ## [1] 3.552

``` r
# since st_prepair performs some preliminary processing to prepare data
# for the prepair tool, we can also print the time spent just cleaning the
# polygons using the prepair tool to measure its performance alone (seconds)
print(attr(y, "prepair_time"))
```

    ## [1] 3.275

``` r
# plot repaired data
plot(st_geometry(y), main = "repaired data")
```

<img src="man/figures/README-cleaned_data-1.png" style="display: block; margin: auto;" />

Next we will repair the data using the [*lwgeom R* package](https://CRAN.R-project.org/package=lwgeom) and record how long this tool takes to repair the data.

``` r
# repair geometries using lwgeom and record processing time
lwgeom_time <- system.time({
  z <- st_make_valid(x)
})[[3]]

# preview cleaned data
print(z)
```

    ## Simple feature collection with 8 features and 1 field
    ## geometry type:  MULTIPOLYGON
    ## dimension:      XY
    ## bbox:           xmin: -179.999 ymin: -89.99983 xmax: 180 ymax: 49.60178
    ## epsg (SRID):    4326
    ## proj4string:    +proj=longlat +datum=WGS84 +no_defs
    ##           ADMIN                       geometry
    ## 12   Antarctica MULTIPOLYGON (((167.1133 -7...
    ## 41        Chile MULTIPOLYGON (((-67.49094 -...
    ## 132     Liberia MULTIPOLYGON (((-9.433115 7...
    ## 174 New Zealand MULTIPOLYGON (((169.1804 -5...
    ## 190     Romania MULTIPOLYGON (((23.32512 43...
    ## 197     Senegal MULTIPOLYGON (((-13.08374 1...
    ## 211    Slovakia MULTIPOLYGON (((19.62695 49...
    ## 225       Tonga MULTIPOLYGON (((-176.207 -2...

``` r
# verify that all geometries are now valid
all(st_is_valid(z))
```

    ## [1] TRUE

``` r
# print processing time (seconds)
print(lwgeom_time)
```

    ## [1] 1.454

As we can see, the `st_prepair` function in this package took 2.1 seconds longer to complete. Therefore do not use this package to clean your data.

### References

[Ledoux H, Arroyo Ohori K, and Meijers M (2014) A triangulation-based approach to automatically repair GIS polygons. Computers & Geosciences 66:121--131](https://doi.org/10.1016/j.cageo.2014.01.009).

### Citation

This is an experimental package. You should not cite this package because you should not be using it.


NEWS.md
# prepairr 0.0.0.3

- Report timings in README based on high performance compiler flags.
- Add `prepair_time` attribute to outputs to _prepair_ tool run time.
- Minor tweaks to package documentation.

# prepairr 0.0.0.2

- Initial developmental release on GitHub.

# prepairr 0.0.0.1

- Initial commit.


R/st_prepair.R
#' Repair geometry using the prepair tool
#'
#' Repair invalid geometry using constrained triangulation via the prepair
#' tool (Ledoux, Arroyo Ohori & Meijers, 2014,
#' \url{https://github.com/tudelft3d/prepair}).
#'
#' @param x \code{\link[sf]{sf}} or \code{\link[sf]{sfc}} object.
#'
#' @param min_area \code{numeric} minimum area permitted. Polygons with an
#'   area below this threshold are treated as slivers and omitted from analysis.
#'   Defaults to 0.00001.
#'
#' @details Note that any non-polygonal geometries (e.g. point or line data) are
#'   not subjected to the data cleaning process.
#'
#' @return Object with repaired geometry. This object also has a
#'   \code{"prepair_time"} attribute the records the time spent cleaning
#'   the data using the prepair tool.
#'
#' @references
#' Ledoux H, Arroyo Ohori K, and Meijers M (2014) A triangulation-based
#' approach to automatically repair GIS polygons. Computers & Geosciences
#' 66:121--131
#'
#' @seealso \code{\link[lwgeom]{st_make_valid}},
#'   \code{\link[sf]{st_is_valid}}.
#'
#' @examples
#' # create sf object with invalid geometry
#' x <- st_sfc(st_polygon(list(rbind(c(0, 0), c(0.5, 0), c(0.5, 0.5),
#'                                   c(0.5, 0), c(1, 0), c(1, 1), c(0, 1),
#'                                   c(0, 0)))))
#'
#' # verify that geometry is invalid
#' suppressWarnings(st_is_valid(x))
#'
#' # fix geometry using prepair
#' y <- st_prepair(x)
#'
#' # verify that repaired geometry is valid
#' st_is_valid(y)
#'
#' # plot the original and repaired geometries
#' par(mfrow = c(1, 2))
#' plot(st_geometry(x), main = "invalid geometry")
#' plot(st_geometry(y), main = "repaired geometry")
#'
#' @export
st_prepair <- function(x, min_area = 1e-5) UseMethod("st_prepair")

#' @export
st_prepair.sf <- function(x, min_area = 1e-5) {
  # assert arguments are valid
  assertthat::assert_that(inherits(x, "sf"), assertthat::is.number(min_area),
                          assertthat::noNA(min_area))
  # repair geometries
  g <- st_prepair.sfc(sf::st_geometry(x))
  sf::st_geometry(x) <- g
  attr(x, "prepair_time") <- attr(g, "prepair_time")
  x
}

#' @export
st_prepair.sfc <- function(x, min_area = 1e-5) {
  # assert arguments are valid
  assertthat::assert_that(inherits(x, "sfc"), assertthat::is.number(min_area),
                          assertthat::noNA(min_area))
  # store attributes
  x_crs <- sf::st_crs(x)
  x_prec <- sf::st_precision(x)
  # replace sliver geometries with empty geometries
  x_pl <- grepl("POLYGON", sf::st_geometry_type(x))
  x <- replace(x, which((as.numeric(sf::st_area(x)) < 1) & x_pl),
               list(sf::st_geometrycollection()))
  # check if any non-polygon objects detected
  if (!all(x_pl))
    warning("non-polygonal geometries will not be repaired", immediate. = TRUE)
  # convert MULTIPOLYGONS to POLYGONS
  x_pl <- grepl("POLYGON", sf::st_geometry_type(x))
  s <- sf::st_sf(id = which(x_pl), geometry = x[x_pl])
  st_crs(s) <- st_crs()
  suppressWarnings({s <- sf::st_cast(s, "POLYGON")})
  # find number of vertices in each polygon so that we can omit
  # polygons with four vertices when running prepair since it can't handle them
  nv <- vapply(s$geometry, function(x) nrow(x[[1]]), numeric(1))
  nv_gte4 <- nv > 4
  # clean polygons using prepair
  s2 <- data.frame(id = s$id[nv_gte4])
  prepair_time <- system.time({
    s2$geometry <- rcpp_prepair(sf::st_as_text(s$geometry[nv_gte4]), min_area)
  })[[3]]
  s2 <- st_as_sf(s2, wkt = "geometry")
  # add in polygons with four vertices, note that we manually set
  # some default CRS to ensure that rbind works
  if (any(nv_gte4)) {
    st_crs(s2) <- st_crs("+init=epsg:4326")
    st_crs(s) <- st_crs("+init=epsg:4326")
    s2 <- rbind(s2, s[nv_gte4, ])
    s2 <- s2[order(s2$id), , drop = FALSE]
  }
  # dissolve polygons to convert fixed polygons to multipolygons
  s2 <- lapply(split(s2$geometry, s2$id), do.call, what = sf::st_union)
  # merge dissolved polygons with original sfc object
  x <- replace(x$geometry, which(x_pl), s2)
  # prepare output
  x <- st_sfc(x)
  st_crs(x) <- x_crs
  st_precision(x) <- x_prec
  attr(x, "prepair_time") <- prepair_time
  # return output
  x
}
